osu.Game/Rulesets/Difficulty/Utils/ObjectLink.cs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace osu.Game.Rulesets.Difficulty.Utils
+{
+    internal class ObjectLink<T> : IReadOnlyList<T>
```

Comment:
At first glance, I was quite confused on what `current.Previous[i]` means (almost felt like it's looking at a previous set of objects and iterating through them from 0 to length), until I read through the usages and checked how this exactly works, to realize that it's an indexer that works in reverse, starting from the object prior to the one specified in this node.

I may be the only one, but I personally would've preferred going for a structure similar to `LinkedList`/`LinkedListNode` in which you would loop through the value of `current.Previous` until it is null or you have reached the desired length, i.e.:
```csharp
var previous = current.Previous;

// loop until object zero
while (previous != null)
{
    // ...
    previous = previous.Previous;
}
// or loop through 25 previous objects at maximum.
for (int i = 0; i < 25 && previous != null; i++)
{
    // ...
    previous = previous.Previous;
}
```

But I'm not fully aware on all the usages of iterating on previous objects, and this may be going one direction behind. Curious to hear more opinions.

If it was decided that we still want `ObjectLink` as-is, then it definitely needs xmldoc to explain how it's supposed to work, and potentially a better name (`LinkedObjectNode` or otherwise).

Reply:
I've we're looking to have full backwards traversal, can't this whole thing be done without a new data type? ie. just using LINQ or otherwise.

We'll need to decide whether the previous history is worthy of extra memory/alloc overhead or lower performance during use (ie. how often this is going to be used going forward). Initial thoughts is that this class should not be required for an initial implementation during experimentation stages (and is potentially early optimisation).

Reply:
First of all I will say that being able to reach N objects back is a common operation in diffcalc and it would be nice to get some sort of first-class syntax support for it, just so that the code reads well.

I've mentioned this offhand previously privately to @apollo-dw but I'm generally not sold on this linked-list-like structure either. Linked lists are only somewhat tolerable at sequential traversal in the link direction and bad at everything else. Traversals more than one item back start to be annoying and expensive.

If the linked list becomes an issue then I proposed that you could have a faux-iterator wrapping a list, wherein you'd have an internal "pointer" to the current item, and be able to look N objects back as far as you like in constant time. But as stated above, maybe this is all just an overoptimisation at this point.

Reply:
Hmm... an iterator wouldn't be overdoing it necessarily: it'd have the additional benefit of retrieving a `Next` object which is pretty valuable. Will be looking into this approach if that sounds agreeable with everyone